### ISSUE TITLE
Impossible to implement async `IntoWs`?

### ISSUE THREAD

**Issue by bspeice** [websockets-rs/rust-websocket#198]
My goal is to build an object-pool `Codec` implementation to avoid allocations for all the frames/messages that come through a websocket. However, despite [server/mod.rs](https://github.com/websockets-rs/rust-websocket/blob/master/src/server/mod.rs#L83) mentioning the `IntoWs` trait, it appears impossible to implement in a separate crate because [`Upgrade` in upgrade/async.rs](https://github.com/websockets-rs/rust-websocket/blob/master/src/server/upgrade/async.rs#L62) is a struct that I can't patch the implementation for. Thus, I'd be able to adapt the underlying stream (so having new exotic things beyond just what `tokio` provides), but the *protocol* details are hidden from other crates.

Am I understanding this correctly? And is that all the more reason for #180?

**Comment by vi** [comment]
Why do you need your own implementation of `IntoWs`?

There is a blanket implementation provided for all `AsyncRead + AsyncWrite + 'static`.

Non-allocating buffers look a non-small change in rust-websocket architecture and is probably not worth pursuing now, given [next-gen rust-websocket](https://github.com/enzious/rust-websocket) may be on the way.

You can also experiment with patching rust-websocket code using [replace section in Cargo.toml](https://doc.rust-lang.org/cargo/reference/manifest.html#the-replace-section).

**Comment by bspeice** [comment]
Non-allocating buffers isn't the issue, the `IntoWs` implementation is because that controls the `Upgrade` object being returned which ultimately determines the `MessageProtocol` used. I want to swap out `MessageProtocol` but in order to use the rest of the infrastructure I have to provide `IntoWs`.
